### PR TITLE
[cmake] Correct and optimize dependencies between cmake targets

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [ ] Update documentation

### CODE REVIEW

**Review by MksmOrlov — COMMENTED** [review]
Tests and run are okay, but wait for the chromium review first

File: sc-kpm/sc-common/CMakeLists.txt [hunk 1]
```diff
@@ -1,20 +1,8 @@
-set(SC_COMMON_SRC ${CMAKE_CURRENT_LIST_DIR})
+file(GLOB_RECURSE SOURCES "*.c" "*.h")
```
Comment (FallenChromium) [inline sc-kpm/sc-common/CMakeLists.txt:1]: Please don't use it. CMake's build system is designed to automatically track dependencies between source files and headers. When you use `file(GLOB_RECURSE ...)`, CMake loses the ability to track these dependencies accurately. This can lead to unnecessary rebuilds of files that haven't actually changed, or it can fail to rebuild files that have changed. There's CONFIGURE_DEPENDS, but it's not used in this script, only supported on CMake 3.12+ and isn't as portable
https://stackoverflow.com/questions/32411963/why-is-cmake-file-glob-evil
https://discourse.cmake.org/t/is-glob-still-considered-harmful-with-configure-depends/808

**Comment by FallenChromium** [comment]
Apart from the `GLOB` usage this PR looks good to me! Thanks for specifying the linking types, it's a very good change in the long-term.

**Comment by FallenChromium** [comment]
Does it make sense to link targets with gcov unconditionally? I think we had a special flag for that

File: sc-kpm/sc-agents-common/CMakeLists.txt [hunk 2]
```diff
@@ -1,14 +1,22 @@
 set(SC_AGENTS_COMMON_SRC "${CMAKE_CURRENT_LIST_DIR}")
 
-file(GLOB_RECURSE SOURCES "*.cpp" "*.hpp")
+file(GLOB SOURCES CONFIGURE_DEPENDS
+    "*.cpp" "*.hpp"
+    "*/*.cpp" "*/*.hpp"
+)
 
-include_directories(${SC_AGENTS_COMMON_SRC} ${SC_MEMORY_SRC} ${SC_KPM_SRC})
+find_package(nlohmann_json 3.2.0 REQUIRED)
```
Comment (FallenChromium) [inline sc-kpm/sc-agents-common/CMakeLists.txt:8]: It probably makes more sense to do this in the root CMakeLists.txt of the sc-machine  (unless sc-kpm is the only module that uses json, but I believe that at least sc-server and sc-ui also use nlohmann_json)

File: sc-memory/sc-core/CMakeLists.txt [hunk 3]
```diff
@@ -1,21 +1,15 @@
-set(SC_CORE_ROOT ${CMAKE_CURRENT_SOURCE_DIR})
+set(SC_CORE_SRC ${CMAKE_CURRENT_SOURCE_DIR})
 
-file(GLOB_RECURSE SOURCES "*.c" "*.h" "*.cpp" "*.hpp")
+file(GLOB SOURCES CONFIGURE_DEPENDS
+    "*.c" "*.h"
+    "*/*.c" "*/*.h"
+    "*/*/*.c" "*/*/*.h"
+    "*/*/*/*.c" "*/*/*/*.h"
+)
```
Comment (FallenChromium) [inline sc-memory/sc-core/CMakeLists.txt:8]: Would've been much nicer if we had CMake targets for sc-core submodules such as sc-store and sc-container, but that's just thinking out loud

File: sc-tools/sc-builder/CMakeLists.txt [hunk 4]
```diff
@@ -3,15 +3,20 @@ configure_file(
...
 add_library(sc-builder-lib SHARED ${SOURCES})
-include_directories(${SC_MEMORY_SRC} ${SC_CONFIG_UTILS_SRC} ${GLIB2_INCLUDE_DIRS})
-target_link_libraries(sc-builder-lib sc-memory sc-config-utils)
+target_link_libraries(sc-builder-lib
+    LINK_PRIVATE sc-memory
+    LINK_PUBLIC sc-config-utils)
+target_include_directories(sc-builder-lib
+    PRIVATE ${SC_MEMORY_SRC}
+    PUBLIC ${SC_CONFIG_UTILS_SRC}
+)
```
Comment (FallenChromium) [inline sc-tools/sc-builder/CMakeLists.txt:16]: Actually I don't get why do we include ${SC_MEMORY_SRC} if all public includes should be propagated why target_link_libraries. Same with config utils - this line should not be needed, nor the variables that are set from other targets

File: sc-tools/sc-machine-runner/CMakeLists.txt [hunk 5]
```diff
@@ -1,14 +1,20 @@
...
-include_directories(${SC_MEMORY_SRC} ${SC_CONFIG_UTILS_SRC} ${GLIB2_INCLUDE_DIRS})
-target_link_libraries(sc-machine-runner sc-memory sc-config-utils)
+target_link_libraries(sc-machine-runner
+    LINK_PRIVATE sc-memory
+    LINK_PRIVATE sc-config-utils
+)
+target_include_directories(sc-machine-runner
+    PUBLIC ${SC_MEMORY_SRC}
+    PUBLIC ${SC_CONFIG_UTILS_SRC}
+)
```
Comment (FallenChromium) [inline sc-tools/sc-machine-runner/CMakeLists.txt:14]: This is confusing. You link sc-memory privately, but make the include directories public. Why?

**Comment by FallenChromium** [comment]
Setting both "target_link_libraries" and "target_include_directories" is a confusing way to set up dependencies and should be avoided. See [this CMake forum post](https://discourse.cmake.org/t/anyway-to-automatically-add-include-dir-when-using-target-link-libraries/2811) for more info

File: sc-memory/sc-core/CMakeLists.txt (hunk 3 above)
Comment (NikitaZotov) [reply sc-memory/sc-core/CMakeLists.txt:8]: It had better to make it later. It's not important now.

File: cmake/codegen.cmake [hunk 6]
```diff
@@ -55,10 +58,12 @@ macro(sc_codegen_ex Target SrcPath OutputPath)
...
         )
 
-        set (SUB_TARGET "${Target}-GenerateCode")
+        set(SUB_TARGET "${Target}-GenerateCode")
         add_custom_target(${SUB_TARGET} DEPENDS ${CACHE_FILE} SOURCES ${HEADER_FILES})
 
-        add_dependencies(${SUB_TARGET} sc-code-generator)
+        if(TARGET sc-code-generator)
+            add_dependencies(${SUB_TARGET} sc-code-generator)
+        endif()
```
Comment (FallenChromium) [inline cmake/codegen.cmake:66]: If possible, I'd like to implement codegen and other similar targets using [ExternalProject_Add](https://cmake.org/cmake/help/latest/module/ExternalProject.html), but that's a subject of discussions and should be implemented in a separate PR anyway

**Review by FallenChromium — APPROVED** [review]
